docs/Dialects/Seq/RationaleSeq.md

```diff
@@ -217,3 +214,99 @@ information.
 - Initial value: this register is uninitialized. Using an uninitialized value
 results in undefined behavior. We will add an `initialValue` attribute if
 this proves insufficient.
+
+## The High-Level Memory Abstraction
+The  `seq.hlmem` (high-level memory operation) intends to capture the semantics
+of a memory which eventually map to some form on-chip resources - whether being
+FPGA or ASIC-based.
+The abstraction aims to abstract away the physical implementation details of the
+memory, and instead focus on the external interface and access semantics of the
+memory. this, in turn, facilitates analysis and transformation (e.g. memory
+merging, read/write conflicts, merging, etc.) and may serve as a target for
+other high-level abstractions.
+
+The high-level memory abstraction is split into two parts:
+* Memory *allocation* is defined by the `seq.hlmem` operation. This operation defines
+the inner workings of the memory. This includes:
+  - number of read/write ports
```

Comment:
Couldn't the number of read/write ports be inferred from the number of `seq.read`/`seq.write` operations? It'd be harder to create invalid IR.

Reply:
Along that same line of thinking, why not attach latency to the port rather than the allocation to better model asymmetric latencies? That would also have the advantage of having the latency closer to the driving logic which has to worry about cycle delays.

Reply:
> Couldn't the number of read/write ports be inferred from the number of seq.read/seq.write operations? It'd be harder to create invalid IR.

And thus have read/write port ops reference the memory symbol rather than the memory generate SSA values? I can see pros/cons for both. As you mention, it will indeed be easier to create valid IR since there is no more any single-use constraint.

If this memory eventually is to evolve into something even higher level, this may make it harder to create the notion of multiple "virtual ports" accessing a single "physical port" (e.g. the abstraction may perform arbitration of a single physical port). But, i see no benefits to having that - it diverges from the abstraction level proposed in this rationale, and it will not be able to conform to the descision to strictly model read/write latencies.

So, i am in favor.

> Along that same line of thinking, why not attach latency to the port rather than the allocation to better model asymmetric latencies? That would also have the advantage of having the latency closer to the driving logic which has to worry about cycle delays.

Wrt. assymetric latencies, is there a need to model asymmetry intra-port type (N read ports with N different read latencies, M write ports with M different latencies). I initially placed latency as a memory-specific constraint under the assumption that all ports of a given direction must have identical latency.

Reply:
> Wrt. assymetric latencies, is there a need to model asymmetry intra-port type (N read ports with N different read latencies, M write ports with M different latencies). I initially placed latency as a memory-specific constraint under the assumption that all ports of a given direction must have identical latency.

I don't know that there is a need, but that's what I was thinking it'd be able to model. A lowering pass which supports only symmetric latencies could use the minimum of all the port latencies, then build delays for the rest to easily support asymmetric latencies.

Reply:
The more I think about it, the more I like allowing per-port latencies as it could be used to relieve timing pressure. For instance, if scheduling ends up with slack, one could just give it to the memory port so that if all the memory accesses have a high slack, the memory optimizer has more flexibility.